### PR TITLE
[PYTHON] Infer calls improvement

### PR DESCRIPTION
**Pull request by jiwaszki**
### Details:
 - New logic for infer calls for `InferRequest` class and it's derivatives. C++ bindings provides two standardized ways of passing inputs: (1) with dictionary of keys and Tensors, (2) single Tensor for one-input models.
 - If passed data origin is of numpy array, copy data to already allocated Tensor.
 - If passed data is of scalar type, create temporary Tensor and set it "in-place".
 - Whole process is being dispatched on Python side. Numpy arrays are not transferred to C++, only Tensor inputs are passed.
 - PR removes code responsible for getting input types (see removed function: `get_input_types`) and relies only on OV Python API itself.
 - **TODO**: add more tests?

### Tickets:
 - *77653*

### CODE REVIEW

File: src/bindings/python/src/pyopenvino/core/async_infer_queue.cpp [hunk 1]
```diff
@@ -175,9 +175,55 @@ void regclass_AsyncInferQueue(py::module m) {
...
+        R"(
+            Run asynchronous inference using the next available InferRequest.
+
+            This function releases the GIL, so another Python thread can
+            work while this function runs in the background.
+
+            :param inputs: Data to set on single input tensor of next available InferRequest from
+            AsyncInferQueue's pool.
+            :type inputs: openvino.runtime.Tensor
+            :param userdata: Any data that will be passed to a callback
```
Comment (p-wysocki) [inline src/bindings/python/src/pyopenvino/core/async_infer_queue.cpp:212]: Shouldn't there be a `:type userdata:`?

Comment (p-wysocki) [inline src/bindings/python/src/pyopenvino/core/async_infer_queue.cpp:212]: Shouldn't there be a `:type userdata:`?

Comment (jiwaszki) [reply src/bindings/python/src/pyopenvino/core/async_infer_queue.cpp:212]: Good catch! I might missed it during refactoring.

**Review by p-wysocki — APPROVED** [review]

File: src/bindings/python/tests/test_inference_engine/test_infer_request.py [hunk 2]
```diff
@@ -45,6 +45,44 @@ def create_simple_request_and_inputs(device):
...
+    request = compiled.create_infer_request()
+    tensor1 = Tensor(ov_type, input_shape)
+    tensor1.data[:] = np.array([6, 7, 8, 9, 0])
+    array1 = np.array([1, 2, 3, 4, 5], dtype=numpy_dtype)
+
+    return request, tensor1, array1
+
+
+def abs_model_with_data(device, ov_type, numpy_dtype):
+    input_shape = [1, 4]
```
Comment (mitruska) [inline src/bindings/python/tests/test_inference_engine/test_infer_request.py:70]: Consider additional test where model has input with dynamic shape.

**Review by mitruska — APPROVED** [review]
Great to see usage of such features as `@singledispatch`! 🚀 
I can see valuable improvements in this PR, 
it would be good to have full validation checks before merging.

File: src/bindings/python/src/pyopenvino/core/async_infer_queue.cpp (hunk 1 above)
Comment (jiwaszki) [reply src/bindings/python/src/pyopenvino/core/async_infer_queue.cpp:212]:
```suggestion
            :param userdata: Any data that will be passed to a callback
            :type userdata: Any
```

**Review by mbencer — APPROVED** [review]

**Pull request merged**